service/matching/tasklist/matcher.go

```diff
@@ -445,32 +438,17 @@ func (tm *TaskMatcher) fwdrAddReqTokenC() <-chan *ForwarderReqToken {
 	return tm.fwdr.AddReqTokenC()
 }
 
-func (tm *TaskMatcher) ratelimit(ctx context.Context) (*rate.Reservation, error) {
```

Comment:
this whole func just plain doesn't make sense originally tbh.
it's re-implementing `Wait` while _using_ `Wait`:
- it's avoiding `Wait` if context is canceled, but `Wait` does that for you.
- it's `Wait`-ing if there is no deadline.  this is fine.
- it's `Reserve`-ing if there is a delay, and sleeping until the reservation's deadline... which is _exactly_ what `Wait` does
- it returns a reservation only after its reserved time has passed, which means canceling will always do nothing

So I just simplified it.
Now it waits.  That either works or it doesn't.  If it doesn't spend time waiting, it returns "throttled" like before.

Reply:
Could you add a unit test for the Wait function to verify that the behavior is the same as before?
As far as I see codecov is not happy with the coverage.

Reply:
Yeah, it seems likely this is just unnecessarily complex, but there's been a few dragons in the tasklist handling already. +1 to the request for a unit test of the behaviour if possible (though tbh I think the scarier problems are more likely in async dispatch)